graphql/util.go

```diff
@@ -771,3 +772,116 @@ func GetMyVolumes(user *user.DBUser) ([]restModel.APIVolume, error) {
+	}
+	vol, err := host.FindVolumeByID(volumeId)
+	if err != nil {
+		return false, http.StatusInternalServerError, InternalServerError, errors.Wrapf(err, "can't get volume '%s'", volumeId)
+	}
+	if vol == nil {
+		return false, http.StatusBadRequest, ResourceNotFound, errors.Errorf("volume '%s' does not exist", volumeId)
+	}
+	mgr, err := getManager(ctx, vol)
+	if err != nil {
+		return false, http.StatusInternalServerError, InternalServerError, err
+	}
+	if hostId == "" {
+		return false, http.StatusBadRequest, InputValidationError, errors.New("must specify host id")
+	}
+	var h *host.Host
+	h, err = host.FindOneId(hostId)
+	if err != nil {
+		return false, http.StatusInternalServerError, InternalServerError, errors.Wrapf(err, "can't get host '%s'", vol.Host)
+	}
+	if h == nil {
+		return false, http.StatusBadRequest, ResourceNotFound, errors.Errorf("host '%s' does not exist", hostId)
+	}
+	err = spawnHostForTestCode(ctx, mgr, h)
+	if err != nil {
+		return false, http.StatusInternalServerError, InternalServerError, err
+	}
+	if vol.AvailabilityZone != h.Zone {
+		return false, http.StatusBadRequest, InputValidationError, errors.New("host and volume must have same availability zone")
+	}
+	if err = mgr.AttachVolume(ctx, h, &host.VolumeAttachment{VolumeID: vol.ID}); err != nil {
+		return false, http.StatusInternalServerError, InternalServerError, errors.Wrapf(err, "can't attach volume '%s'", vol.ID)
+	}
+	return true, http.StatusOK, "", nil
+}
+
+func DetachVolume(ctx context.Context, volumeId string) (bool, int, GqlError, error) {
+	if volumeId == "" {
+		return false, http.StatusBadRequest, InputValidationError, errors.New("must specify volume id")
+	}
+	vol, err := host.FindVolumeByID(volumeId)
+	if err != nil {
+		return false, http.StatusInternalServerError, InternalServerError, errors.Wrapf(err, "can't get volume '%s'", volumeId)
+	}
+	if vol == nil {
+		return false, http.StatusBadRequest, ResourceNotFound, errors.Errorf("volume '%s' does not exist", volumeId)
+	}
+	mgr, err := getManager(ctx, vol)
+	if err != nil {
+		return false, http.StatusInternalServerError, InternalServerError, err
+	}
+	if vol.Host == "" {
+		return false, http.StatusBadRequest, InputValidationError, errors.Errorf("volume '%s' is not attached", vol.ID)
+	}
+	h, err := host.FindOneId(vol.Host)
+	if err != nil {
+		return false, http.StatusInternalServerError, InternalServerError, errors.Wrapf(err, "can't get host '%s' for volume '%s'", vol.Host, vol.ID)
+	}
+	if h == nil {
+		if err = host.UnsetVolumeHost(vol.ID); err != nil {
+			grip.Error(message.WrapError(err, message.Fields{
+				"message": fmt.Sprintf("can't clear host '%s' from volume '%s'", vol.Host, vol.ID),
+				"route":   "graphql/util",
+				"action":  "DetachVolume",
+			}))
+		}
+		return false, http.StatusInternalServerError, InternalServerError, errors.Errorf("host '%s' for volume '%s' doesn't exist", vol.Host, vol.ID)
+	}
+	err = spawnHostForTestCode(ctx, mgr, h)
```

Comment:
Instead of doing this, could set up the mock host at the beginning of the test? The test code right now doesn't quite have the flexibility to have individual tests specify their setup code, so we could either add that, or just always setup some hosts to use like we do with the user